### PR TITLE
new endpoint for editing service account

### LINKED ISSUE #3079: Create new endpoint for editing SA
**User Story**
Create new endpoint to be able to edit SA for the project
**Acceptance criteria**
The user can change SA name and group

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**: new endpoint for editing service account

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3079

```release-note
new endpoint for editing service account for the project: PUT /api/v1/projects/{project_id}/serviceaccounts/{serviceaccount_id}
```

### CODE REVIEW

File: api/pkg/handler/routes_v1.go [hunk 1]
```diff
@@ -267,6 +267,10 @@ func (r Routing) RegisterV1(mux *mux.Router, metrics common.ServerMetrics) {
 		Path("/projects/{project_id}/serviceaccounts").
 		Handler(r.listServiceAccounts())
 
+	mux.Methods(http.MethodPut).
+		Path("/projects/{project_id}/serviceaccounts/{serviceaccount_id}").
+		Handler(r.editServiceAccount())
```
Comment (p0lyn0mial) [inline api/pkg/handler/routes_v1.go]: please change the name to `updateServiceAccount`

File: api/pkg/handler/v1/serviceaccount/sa.go [hunk 2]
```diff
@@ -107,6 +108,56 @@ func ListEndpoint(projectProvider provider.ProjectProvider, serviceAccountProvid
 	}
 }
 
+// EditEndpoint changes the service account group and/or name in the given project
+func EditEndpoint(projectProvider provider.ProjectProvider, serviceAccountProvider provider.ServiceAccountProvider) endpoint.Endpoint {
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/serviceaccount/sa.go]: Please change the name to `UpdateEndpoint`, for example https://github.com/kubermatic/kubermatic/blob/master/api/pkg/handler/v1/project/project.go#L130

File: api/pkg/handler/v1/serviceaccount/sa.go [hunk 3]
```diff
@@ -115,6 +185,31 @@ type addReq struct {
...
+// idReq represents a request that contains service account ID in the path
+type idReq struct {
+	// in: path
+	ServiceAccountID string `json:"serviceaccount_id"`
+}
+
+// editReq defines HTTP request for editServiceAccount
+// swagger:parameters editServiceAccount
+type editReq struct {
+	addReq
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/serviceaccount/sa.go:188]: should we also change "Update a project" EP to accept `apiv1.Project` in the body ?
https://github.com/kubermatic/kubermatic/blob/master/api/pkg/handler/v1/project/project.go#L230

File: api/pkg/provider/kubernetes/sa.go [hunk 4]
```diff
@@ -129,3 +130,49 @@ func (p *ServiceAccountProvider) List(userInfo *provider.UserInfo, project *kube
...
+// Get method returns service account with given name
+func (p *ServiceAccountProvider) Get(userInfo *provider.UserInfo, name string) (*kubermaticv1.User, error) {
+	if userInfo == nil {
+		return nil, kerrors.NewBadRequest("userInfo cannot be nil")
+	}
+	if len(name) == 0 {
+		return nil, kerrors.NewBadRequest("service account name cannot be empty")
+	}
+
+	serviceAccount, err := p.serviceAccountLister.Get(name)
```
Comment (p0lyn0mial) [inline api/pkg/provider/kubernetes/sa.go]: don't get the resource from the lister, use masterImpersonatedClient instead.

File: api/pkg/handler/v1/serviceaccount/sa.go [hunk 5]
```diff
@@ -107,6 +108,77 @@ func ListEndpoint(projectProvider provider.ProjectProvider, serviceAccountProvid
...
+			}
+
+			if len(existingSAList) > 0 {
+				return nil, errors.NewAlreadyExists("service account", saFromRequest.Name)
+			}
+			sa.Spec.Name = saFromRequest.Name
+
+		}
+
+		bindings, err := memberMapper.MappingsFor(sa.Spec.Email)
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/serviceaccount/sa.go]: use `MapUserToGroup` instead.

File: api/pkg/handler/routes_v1.go [hunk 6]
```diff
@@ -1348,6 +1352,34 @@ func (r Routing) listServiceAccounts() http.Handler {
 	)
 }
 
+// swagger:route PUT /api/v1/projects/{project_id}/serviceaccounts/{serviceaccount_id} serviceaccounts editServiceAccount
+//
+//     Changes membership of the given service account for the given project
```
Comment (p0lyn0mial) [inline api/pkg/handler/routes_v1.go]: please change the description, this EP updates the resource not only changes the membership.

File: api/pkg/handler/v1/serviceaccount/sa.go [hunk 7]
```diff
@@ -107,6 +108,77 @@ func ListEndpoint(projectProvider provider.ProjectProvider, serviceAccountProvid
 	}
 }
 
+// EditEndpoint changes the service account group and/or name in the given project
+func EditEndpoint(projectProvider provider.ProjectProvider, serviceAccountProvider provider.ServiceAccountProvider, memberMapper provider.ProjectMemberMapper) endpoint.Endpoint {
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/serviceaccount/sa.go]: please change to `UpdateEndpoint`, for example https://github.com/kubermatic/kubermatic/blob/master/api/pkg/handler/v1/project/project.go#L130

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]

File: api/pkg/handler/routes_v1.go (hunk 1 above)
Comment (zreigz) [reply api/pkg/handler/routes_v1.go]: done

File: api/pkg/handler/v1/serviceaccount/sa.go (hunk 2 above)
Comment (zreigz) [reply api/pkg/handler/v1/serviceaccount/sa.go]: done

File: api/pkg/handler/v1/serviceaccount/sa.go (hunk 3 above)
Comment (zreigz) [reply api/pkg/handler/v1/serviceaccount/sa.go:188]: yes, I've created the issue for it: #3087

File: api/pkg/provider/kubernetes/sa.go (hunk 4 above)
Comment (zreigz) [reply api/pkg/provider/kubernetes/sa.go]: done

File: api/pkg/handler/v1/serviceaccount/sa.go (hunk 5 above)
Comment (zreigz) [reply api/pkg/handler/v1/serviceaccount/sa.go]: done

File: api/pkg/handler/routes_v1.go (hunk 6 above)
Comment (zreigz) [reply api/pkg/handler/routes_v1.go]: done

File: api/pkg/handler/v1/serviceaccount/sa.go (hunk 7 above)
Comment (zreigz) [reply api/pkg/handler/v1/serviceaccount/sa.go]: done

**Comment by zreigz** [comment]
@p0lyn0mial PTAL

**Pull request merged**